### ISSUE TITLE
[BUG/ISSUE] State_Chm%nDryDep includes non-advected species

### ISSUE THREAD

**Issue by lizziel** [geoschem/geos-chem#1095]
I discovered a bug in GEOS-Chem dry deposition. There is a mismatch in number of dry deposited species stored in `State_Chm%nDryDep` versus number of dry deposited species stored in drydep_mod.F90 module variable `NUMDEP`. More specifically, `NUMDEP` is limited to advected species only while `State_Chm%nDryDep` includes both advected and non-advected species.

A bit of background:
`NUMDEP` is computed in subroutine `INIT_DRYDEP` by looping over advected species (`State_Chm%nAdvect`) and incrementing by one for every species listed as a dry deposited species in the species database. `NUMDEP` is therefore the number of advected dry deposited species. This is what is looped over when printing list of dry deposited species to the log:

```
INIT_DRYDEP: List of dry deposition species:

  #   Name      Species DEPVEL Henry's    React.   Molec.   Aerosol?
                Number Index  Law Const  Factor   Weight   (T or F)
----------------------------------------------------------------------
  1   ACET         1     1     1.0E+05    1.000    0.058     F
  2   ACTA         2     2     4.1E+03    1.000    0.060     F
...
160   TSOG2      217   160     1.0E+05    0.000    0.150     F
161   TSOG3      218   161     1.0E+05    0.000    0.150     F
```

`State_Chm%nDryDep` is computed in subroutine `Init_Species_Database` and is also incremented by one for every species listed as a dry deposited species in the species database. However, this occurs in a loop over all species, both advected and KPP species. In my testrun this value is 174 not 161.

I printed out the mapping between species ID, dry deposition ID in the species database, and dry deposition ID stored in drydep_mod.F90 module variable `NDVZIND`. When looping over `State_Chm%nDryDep` the `NDVZIND` values for advected species match the dry deposition IDs stored in the species database. For non-advected species the dry deposition IDs do not match; NDVZIND is zero for non-advected species. This is shown below, with D, NDVZ, and N corresponding to dry dep ID, value in `NDVZIND`, and species ID:

```
D, NDVZ, N:          160         160         217
D, NDVZ, N:          161         161         218
D, NDVZ, N:          162           0         220
D, NDVZ, N:          163           0         250
D, NDVZ, N:          164           0         251
D, NDVZ, N:          165           0         253
D, NDVZ, N:          166           0         254
D, NDVZ, N:          167           0         255
D, NDVZ, N:          168           0         257
D, NDVZ, N:          169           0         259
D, NDVZ, N:          170           0         260
D, NDVZ, N:          171           0         263
D, NDVZ, N:          172           0         267
D, NDVZ, N:          173           0         271
D, NDVZ, N:          174           0         291
```

The number of advected species in my run is 219, as shown in the log, and advected species always appear first in the species list. `N` values of 220 and above are thus non-advected species.

```
ADVECTED SPECIES MENU
------------------------------------------------
  #  Species Name
  1  ACET
  2  ACTA
...
218  TSOG3
219  XYLE
```

This is a problem for a few reasons. 

- Memory is allocated for several arrays using `State_Chm%nDryDep`, including diagnostics. By including non-advected species these arrays are nearly 10% larger than they need to be. 
- There is inconsistency in looping over `State_Chm%nDryDep` versus `NUMDEP`, using the former to compute dry deposition frequencies and velocities as well as apply dry deposition velocities, and the latter to compute dry deposition velocities using a resistance-in-series model.
- `NDVZIND` is used to get indexes for `State_Chm%DryDepVel` and other variables when setting values. The index mapping is wrong for the case of zeros. Fortran allows zero indexing so this has gone under the radar.

I think the simple fix is to only increment `State_Chm%nDryDep` in `Init_Species_Database` for species that are advected. There may be other values we need to do this for as well. We can put this fix into 13.4.

**Comment by djxjacob** [comment]
Hi Lizzie, nice detective work! Will this have a significant effect on the dry deposition velocities calculated by the model?  Daniel

From: Lizzie Lundgren ***@***.***>
Sent: Friday, January 14, 2022 8:59 PM
To: geoschem/geos-chem ***@***.***>
Cc: Subscribed ***@***.***>
Subject: [geoschem/geos-chem] [BUG/ISSUE] State_Chm%nDryDep includes non-advected species (Issue #1095)


I discovered a bug in GEOS-Chem dry deposition. There is a mismatch in number of dry deposited species stored in State_Chm%nDryDep versus number of dry deposited species stored in drydep_mod.F90 module variable NUMDEP. More specifically, NUMDEP is limited to advected species only while State_Chm%nDryDep includes both advected and non-advected species.

A bit of background:
NUMDEP is computed in subroutine INIT_DRYDEP by looping over advected species (State_Chm%nAdvect) and incrementing by one for every species listed as a dry deposited species in the species database. NUMDEP is therefore the number of advected dry deposited species. This is what is looped over when printing list of dry deposited species to the log:

INIT_DRYDEP: List of dry deposition species:



  #   Name      Species DEPVEL Henry's    React.   Molec.   Aerosol?

                Number Index  Law Const  Factor   Weight   (T or F)

----------------------------------------------------------------------

  1   ACET         1     1     1.0E+05    1.000    0.058     F

  2   ACTA         2     2     4.1E+03    1.000    0.060     F

...

160   TSOG2      217   160     1.0E+05    0.000    0.150     F

161   TSOG3      218   161     1.0E+05    0.000    0.150     F

State_Chm%nDryDep is computed in subroutine Init_Species_Database and is also incremented by one for every species listed as a dry deposited species in the species database. However, this occurs in a loop over all species, both advected and KPP species. In my testrun this value is 174 not 161.

I printed out the mapping between species ID, dry deposition ID in the species database, and dry deposition ID stored in drydep_mod.F90 module variable NDVZIND. When looping over State_Chm%nDryDep the NDVZIND values for advected species match the dry deposition IDs stored in the species database. For non-advected species the dry deposition IDs do not match; NDVZIND is zero for non-advected species. This is shown below, with D, NDVZ, and N corresponding to dry dep ID, value in NDVZIND, and species ID:

D, NDVZ, N:          160         160         217

D, NDVZ, N:          161         161         218

D, NDVZ, N:          162           0         220

D, NDVZ, N:          163           0         250

D, NDVZ, N:          164           0         251

D, NDVZ, N:          165           0         253

D, NDVZ, N:          166           0         254

D, NDVZ, N:          167           0         255

D, NDVZ, N:          168           0         257

D, NDVZ, N:          169           0         259

D, NDVZ, N:          170           0         260

D, NDVZ, N:          171           0         263

D, NDVZ, N:          172           0         267

D, NDVZ, N:          173           0         271

D, NDVZ, N:          174           0         291

The number of advected species in my run is 219, as shown in the log, and advected species always appear first in the species list. N values of 220 and above are thus non-advected species.

ADVECTED SPECIES MENU

------------------------------------------------

  #  Species Name

  1  ACET

  2  ACTA

...

218  TSOG3

219  XYLE

This is a problem for a few reasons.

  *   Memory is allocated for several arrays using State_Chm%nDryDep, including diagnostics. By including non-advected species these arrays are nearly 10% larger than they need to be.
  *   There is inconsistency in looping over State_Chm%nDryDep versus NUMDEP, using the former to compute dry deposition frequencies and velocities as well as apply dry deposition velocities, and the latter to compute dry deposition velocities using a resistance-in-series model.
  *   NDVZIND is used to get indexes for State_Chm%DryDepVel and other variables when setting values. The index mapping is wrong for the case of zeros. Fortran allows zero indexing so this has gone under the radar.

I think the simple fix is to only increment State_Chm%nDryDep in Init_Species_Database for species that are advected. There may be other values we need to do this for as well. We can put this fix into 13.4.

—
Reply to this email directly, view it on GitHub<https://urldefense.proofpoint.com/v2/url?u=https-3A__github.com_geoschem_geos-2Dchem_issues_1095&d=DwMCaQ&c=WO-RGvefibhHBZq3fL85hQ&r=rhdjeCmvErplnjGEj5OoH_CDkPdqB-BDUxqwWC1AzIs&m=o1ZPo59QGpnXrw3SxKTZcBZVQI_3HwYcRa--w4OBVwpn13_OVv0_CpbUPA9eGWpA&s=2SUebZYlLHwmgNpTvjpG5YzUkLOknI0FUtWyShJf3_8&e=>, or unsubscribe<https://urldefense.proofpoint.com/v2/url?u=https-3A__github.com_notifications_unsubscribe-2Dauth_AKHHSVJVGCBAWQBN542BEUTUWDIHFANCNFSM5MAH5SNQ&d=DwMCaQ&c=WO-RGvefibhHBZq3fL85hQ&r=rhdjeCmvErplnjGEj5OoH_CDkPdqB-BDUxqwWC1AzIs&m=o1ZPo59QGpnXrw3SxKTZcBZVQI_3HwYcRa--w4OBVwpn13_OVv0_CpbUPA9eGWpA&s=JeyYe5BTd6B_xcg90tDQzvaR3i99SzE0sWlNHEtoYEY&e=>.
Triage notifications on the go with GitHub Mobile for iOS<https://urldefense.proofpoint.com/v2/url?u=https-3A__apps.apple.com_app_apple-2Dstore_id1477376905-3Fct-3Dnotification-2Demail-26mt-3D8-26pt-3D524675&d=DwMCaQ&c=WO-RGvefibhHBZq3fL85hQ&r=rhdjeCmvErplnjGEj5OoH_CDkPdqB-BDUxqwWC1AzIs&m=o1ZPo59QGpnXrw3SxKTZcBZVQI_3HwYcRa--w4OBVwpn13_OVv0_CpbUPA9eGWpA&s=YlrXJ3bCDbWn7gBMOcpY8mA9ISasnYEaMAj0QL4O2SE&e=> or Android<https://urldefense.proofpoint.com/v2/url?u=https-3A__play.google.com_store_apps_details-3Fid-3Dcom.github.android-26referrer-3Dutm-5Fcampaign-253Dnotification-2Demail-2526utm-5Fmedium-253Demail-2526utm-5Fsource-253Dgithub&d=DwMCaQ&c=WO-RGvefibhHBZq3fL85hQ&r=rhdjeCmvErplnjGEj5OoH_CDkPdqB-BDUxqwWC1AzIs&m=o1ZPo59QGpnXrw3SxKTZcBZVQI_3HwYcRa--w4OBVwpn13_OVv0_CpbUPA9eGWpA&s=WRS2D456Sl1HidZanlxR0jgQWdA328ClpolSSwLMpIw&e=>.

**Comment by lizziel** [comment]
A fix for this is included in PR https://github.com/geoschem/geos-chem/pull/1088 (GEOS-Chem 13.3.4 in GEOSgcm) which is currently under review. See commit https://github.com/geoschem/geos-chem/pull/1088/commits/621268c249495f4a544142483f04a8f9114702f2. If the GEOS updates are not merged in before the 13.4 freeze date on 1/21 then I will put in a fix for this bug separately from that update. It is included as part of the GEOS updates because it fixes a segmentation fault encountered in GEOS only.

**Comment by yantosca** [comment]
PR #1088 was merged into the 13.4.0 development stream, so we can close this issue.  For further discussion, please see #1099
